lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationServerPlatform.h

```diff
@@ -25,90 +25,30 @@ namespace process_gdb_remote {
-
-    // Add a port to the map. If it is already in the map do not modify
-    // its mapping. (used ports remain used, new ports start as free)
-    void AllowPort(uint16_t port);
-
-    // If we are using a port map where we can only use certain ports,
-    // get the next available port.
-    //
-    // If we are using a port map and we are out of ports, return an error.
-    //
-    // If we aren't using a port map, return 0 to indicate we should bind to
-    // port 0 and then figure out which port we used.
-    llvm::Expected<uint16_t> GetNextAvailablePort();
-
-    // Tie a port to a process ID. Returns false if the port is not in the port
-    // map. If the port is already in use it will be moved to the given pid.
-    // FIXME: This is and GetNextAvailablePort make create a race condition if
-    // the portmap is shared between processes.
-    bool AssociatePortWithProcess(uint16_t port, lldb::pid_t pid);
-
-    // Free the given port. Returns false if the port is not in the map.
-    bool FreePort(uint16_t port);
-
-    // Free the port associated with the given pid. Returns false if there is
-    // no port associated with the pid.
-    bool FreePortForProcess(lldb::pid_t pid);
-
-    // Returns true if there are no ports in the map, regardless of the state
-    // of those ports. Meaning a map with 1 used port is not empty.
-    bool empty() const;
-
-  private:
-    std::map<uint16_t, lldb::pid_t> m_port_map;
-  };
-
   GDBRemoteCommunicationServerPlatform(
-      const Socket::SocketProtocol socket_protocol, const char *socket_scheme);
+      const Socket::SocketProtocol socket_protocol, uint16_t gdbserver_port);
 
   ~GDBRemoteCommunicationServerPlatform() override;
 
   Status LaunchProcess() override;
 
-  // Set both ports to zero to let the platform automatically bind to
-  // a port chosen by the OS.
-  void SetPortMap(PortMap &&port_map);
-
   void SetPortOffset(uint16_t port_offset);
 
   void SetInferiorArguments(const lldb_private::Args &args);
 
-  // Set port if you want to use a specific port number.
-  // Otherwise port will be set to the port that was chosen for you.
-  Status LaunchGDBServer(const lldb_private::Args &args, std::string hostname,
-                         lldb::pid_t &pid, std::optional<uint16_t> &port,
-                         std::string &socket_name);
+  Status LaunchGDBServer(const lldb_private::Args &args, lldb::pid_t &pid,
+                         std::string &socket_name, shared_fd_t fd);
 
-  void SetPendingGdbServer(lldb::pid_t pid, uint16_t port,
-                           const std::string &socket_name);
+  void SetPendingGdbServer(const std::string &socket_name);
 
 protected:
   const Socket::SocketProtocol m_socket_protocol;
-  const std::string m_socket_scheme;
-  std::recursive_mutex m_spawned_pids_mutex;
-  std::set<lldb::pid_t> m_spawned_pids;
+  static std::mutex g_spawned_pids_mutex;
+  static std::set<lldb::pid_t> g_spawned_pids;
```

Comment:
I guess this is related to #101475, but I think we should figure out a solution that does not require making these static.

Reply:
There is no a safe way except static. 
Note GDBRemoteCommunicationServerPlatform works in a separated single thread process. I think it is ok.